src/Symfony/Bundle/FrameworkBundle/Secrets/SodiumVault.php

```diff
@@ -0,0 +1,164 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Secrets;
+
+/**
+ * @author Tobias Schultze <http://tobion.de>
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class SodiumVault
+{
+    private $encryptionKey;
+    private $decryptionKey;
+    private $secretsDir;
+    private $encryptionKeyName;
+    private $decryptionKeyName;
+
+    public function __construct(string $secretsDir)
+    {
+        if (!\function_exists('sodium_crypto_box_seal')) {
+            throw new \LogicException('The "sodium" PHP extension is required to deal with secrets. Alternatively, try running "composer require paragonie/sodium_compat" if you cannot enable the extension."');
+        }
+
+        if (!is_dir($secretsDir) && !@mkdir($secretsDir, 0777, true) && !is_dir($secretsDir)) {
+            throw new \RuntimeException(sprintf('Unable to create the secrets directory (%s)', $secretsDir));
+        }
+
+        $this->secretsDir = rtrim($secretsDir, '/'.\DIRECTORY_SEPARATOR).\DIRECTORY_SEPARATOR;
+
+        $baseName = basename($this->secretsDir);
+        $this->encryptionKeyName = $baseName.'.sodium.public.key';
```

Comment:
if you rename `encryption` to `public`, then we should change the messages in commands